common/io_lock.c

```diff
@@ -0,0 +1,74 @@
+#include "io_lock.h"
+#include <assert.h>
+#include <ccan/io/io_plan.h>
+
+struct io_lock {
+	bool free;
+};
+
+/* Struct to hold information while we wait for the lock to be freed */
+struct io_lock_waiter {
+	struct io_plan *(*next)(struct io_conn *conn, void *next_arg);
+	void *arg;
+	struct io_lock *lock;
+	enum io_direction dir;
+};
+
+struct io_lock *io_lock_new(const tal_t *ctx)
+{
+	struct io_lock *lock = tal(ctx, struct io_lock);
+	lock->free = true;
+	return lock;
+}
+
+static struct io_plan *io_lock_try_acquire(struct io_conn *conn,
+					   struct io_lock_waiter *waiter)
+{
+	if (waiter->lock->free) {
+		waiter->lock->free = false;
+		return io_out_always(conn, waiter->next, waiter->arg);
+	} else {
+		return io_out_wait(conn, waiter->lock, io_lock_try_acquire, waiter);
+	}
+}
+
+static struct io_plan *io_lock_acquire_dir(
+    struct io_conn *conn, struct io_lock *lock, enum io_direction dir,
+    struct io_plan *(*next)(struct io_conn *, void *), void *arg)
+{
+	if (lock->free) {
+		lock->free = false;
+		if (dir == IO_OUT)
+			return io_out_always(conn, next, arg);
+		else
+			return next(conn, arg);
+	}
```

Comment:
Is this fast-path a premature optimization?  It avoids the allocation, but I'd write it simple and put a FIXME suggestion in, so we look clever if profiling later reveals a bottleneck.

Reply:
Sounds good :+1: